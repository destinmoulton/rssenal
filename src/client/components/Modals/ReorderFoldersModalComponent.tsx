import * as React from "react";

import { Button, Header, Modal, Popup } from "semantic-ui-react";

import { SortableFolderList } from "./SortableComponents";

import { propertyComparator } from "../../lib/sort";

import { IFolder, TFolders } from "../../types";

export interface IReorderFoldersMapState {
    folders: TFolders;
}

export interface IReorderFolderMapDispatch {
    beginReorderFolders: (foldersArr: IFolder[]) => void;
}

type TAllProps = IReorderFolderMapDispatch & IReorderFoldersMapState;

interface IReorderFoldersModalState {
    foldersAsArray: IFolder[];
    isModalOpen: boolean;
}

class ReorderFoldersModalComponent extends React.Component<
    TAllProps,
    IReorderFoldersModalState
> {
    state: IReorderFoldersModalState = {
        foldersAsArray: [],
        isModalOpen: false
    };

    static getDerivedStateFromProps(props: TAllProps) {
        let tmpArray = props.folders
            .toArray()
            .sort((a, b) => propertyComparator(a, b, "asc", "order", false));
        //Remove the "Uncategorized" folder
        tmpArray.pop();

        return {
            foldersAsArray: tmpArray
        };
    }

    _handleCloseModal = () => {
        this.setState({
            isModalOpen: false
        });
    };

    _handleOpenModal = () => {
        this.setState({
            isModalOpen: true
        });
    };

    _handlePressOK = () => {
        const { foldersAsArray } = this.state;
        const orderedFolders = foldersAsArray.map((folder, index) => {
            folder.order = index + 1;
            return folder;
        });
        this.props.beginReorderFolders(orderedFolders);
        this._handleCloseModal();
    };

    _onSortEnd = (reorderedObj: any) => {
        const newFoldersArray = this._reorderFoldersArray(
            reorderedObj.oldIndex,
            reorderedObj.newIndex
        );
        this.setState({
            foldersAsArray: newFoldersArray
        });
    };

    _reorderFoldersArray(previousIndex: number, newIndex: number): IFolder[] {
        const array = this.state.foldersAsArray.slice(0);
        if (newIndex >= array.length) {
            let k = newIndex - array.length;
            while (k-- + 1) {
                array.push(undefined);
            }
        }
        array.splice(newIndex, 0, array.splice(previousIndex, 1)[0]);
        return array;
    }

    render() {
        const { foldersAsArray, isModalOpen } = this.state;
        return (
            <span>
                <Popup
                    trigger={
                        <Button
                            className="rss-leftmenu-button-reorder"
                            icon="numbered list"
                            onClick={this._handleOpenModal}
                            size="mini"
                        />
                    }
                    content="Reorder Folders"
                />
                <Modal
                    open={isModalOpen}
                    onClose={this._handleCloseModal}
                    size="tiny"
                    className="rss-modal"
                >
                    <Header icon="numbered list" content="Reorder Folders" />
                    <Modal.Content>
                        <SortableFolderList
                            items={foldersAsArray}
                            onSortEnd={this._onSortEnd}
                        />
                    </Modal.Content>
                    <Modal.Actions>
                        <div>
                            <Button
                                icon="cancel"
                                color="orange"
                                floated="left"
                                inverted
                                content="Cancel"
                                onClick={this._handleCloseModal}
                            />
                            <Button
                                icon="checkmark"
                                inverted
                                color="green"
                                content="Reorder"
                                onClick={this._handlePressOK}
                            />
                        </div>
                    </Modal.Actions>
                </Modal>
            </span>
        );
    }
}

export default ReorderFoldersModalComponent;
