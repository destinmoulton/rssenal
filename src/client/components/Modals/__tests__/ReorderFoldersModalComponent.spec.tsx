import * as React from "react";

import { shallow } from "enzyme";

import ReorderFoldersModalComponent from "../ReorderFoldersModalComponent";

import folders from "../../../../../test/data/immutable/folders";

describe("<ReorderFoldersModalComponent />", () => {
    it("renders and matches the snapshot", () => {
        const wrapper = shallow(
            <ReorderFoldersModalComponent
                folders={folders}
                foldersReorder={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
