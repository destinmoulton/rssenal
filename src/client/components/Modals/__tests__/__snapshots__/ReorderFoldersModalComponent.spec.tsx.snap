// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReorderFoldersModalComponent /> renders and matches the snapshot 1`] = `
<span>
  <Popup
    content="Reorder Folders"
    keepInViewPort={true}
    on="hover"
    position="top left"
    trigger={
      <Button
        as="button"
        className="rss-leftmenu-button-reorder"
        icon="numbered list"
        onClick={[Function]}
        role="button"
        size="mini"
      />
    }
  />
  <Modal
    centered={true}
    className="rss-modal"
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    onClose={[Function]}
    open={false}
    size="tiny"
  >
    <Header
      content="Reorder Folders"
      icon="numbered list"
    />
    <ModalContent>
      <sortableList
        axis="y"
        distance={0}
        getHelperDimensions={[Function]}
        hideSortableGhost={true}
        items={
          Array [
            Object {
              "__v": 0,
              "_id": "5b33c4bab2438d5708dc1967",
              "name": "Random",
              "order": 0,
            },
          ]
        }
        lockOffset="50%"
        lockToContainerEdges={false}
        onSortEnd={[Function]}
        pressDelay={0}
        pressThreshold={5}
        shouldCancelStart={[Function]}
        transitionDuration={300}
        useWindowAsScrollContainer={false}
      />
    </ModalContent>
    <ModalActions>
      <div>
        <Button
          as="button"
          color="orange"
          content="Cancel"
          floated="left"
          icon="cancel"
          inverted={true}
          onClick={[Function]}
          role="button"
        />
        <Button
          as="button"
          color="green"
          content="Reorder"
          icon="checkmark"
          inverted={true}
          onClick={[Function]}
          role="button"
        />
      </div>
    </ModalActions>
  </Modal>
</span>
`;
